src/collector/otlp/service.rs

```diff
@@ -376,3 +361,80 @@ fn process_sample(
 
     Ok(())
 }
+
+fn collect_frame_list<V>(
+    loc_start: usize,
+    loc_len: usize,
+    loc_mapping: &Vec<V>,
+    profile_location_indices: &Vec<i32>,
+) -> Result<Vec<V>, Status>
+where
+    V: Copy,
+{
+    let loc_rng = loc_start..loc_start.saturating_add(loc_len);
+
+    // Collect frame list.
+    let mut frame_list = Vec::with_capacity(loc_rng.len().min(128));
+    for loc_index in loc_rng {
+        let Some(location_table_idx) = profile_location_indices.get(loc_index as usize) else {
+            return Err(Status::invalid_argument(
+                "location_indices: index is out of bounds",
+            ));
+        };
+        let Some(frame) = loc_mapping.get(*location_table_idx as usize) else {
+            return Err(Status::invalid_argument(
+                "location_table: index is out of bounds",
+            ));
+        };
+        frame_list.push(*frame);
+    }
+
+    return Ok(frame_list);
+}
+
+#[cfg(test)]
+mod tests {
+    use itertools::Itertools;
+
+    use super::*;
+
+    #[test]
+    fn sample_frame_list() -> Result<(), Status> {
+        let loc_mapping = (0..11).collect_vec();
+        let location_indices = vec![4, 9, 6, 2, 7, 4, 4, 2, 0, 1, 2, 3, 5];
+
+        assert_eq!(
```

Comment:
Add comments/names to the test about what they are testing, e.g.
"2 items from start", "0 items" or "all items"

Add two more tests: "2 items from end", "2 item from middle"

Reply:
Not a rust guy here... maybe adding the "names" as error message to the assert_eq macros is better!?

Reply:
Something like this? 14198b81ef3f4b01da5a139ef16ed7b638bc3bfb

Reply:
:+1: